config-provisioning/src/main/java/com/yahoo/config/provision/CapacityPolicies.java

```diff
@@ -144,13 +155,53 @@ private NodeResources defaultResources(ClusterSpec clusterSpec) {
         }
     }
 
-    private NodeResources clusterControllerResources(ClusterSpec clusterSpec, Architecture architecture) {
+    private NodeResources clusterControllerResources(ClusterSpec clusterSpec, Architecture architecture, long contentNodes) {
         // 1.32 fits floor(8/1.32) = 6 cluster controllers on each 8Gb host, and each will have
         // 1.32-(0.7+0.6)*(1.32/8) = 1.1 Gb real memory given current taxes.
         var memory = architecture == x86_64
                 ? tuning.clusterControllerMem(1.32)
                 : tuning.clusterControllerMem(1.50);
-        return versioned(clusterSpec, Map.of(new Version(0), new NodeResources(0.25, memory, 10, 0.3)));
+
+        var adjustedMemory = adjustClusterControllerMemory(memory, contentNodes);
+        // But go back to use overridden memory if set (through feature flag)
+        if (tuning.clusterControllerMemoryGiB() > 0.0) {
+            adjustedMemory = memory;
+        }
+
+        return versioned(clusterSpec, Map.of(new Version(0), new NodeResources(0.25, memory, 10, 0.3),
+                                             new Version(8, 574, 5), new NodeResources(0.25, adjustedMemory, 10, 0.3)));
+    }
+
+    // Adjust memory based on number of content nodes in all content clusters
+    // Note: nodeCount is 0 if unknown.
+    private static double adjustClusterControllerMemory(double memory, long nodeCount) {
+        int count = (int) nodeCount;
+        // We have seen clusters with ~100 nodes needing at least 1.6 GiB on x86_64
+        // Adjust memory based on number of content nodes (which is a simple way to model the O(n^2) behavior
+        // of the increase in communication between cluster controller and nodes (and thus memory)).
+        // Increase in steps to avoid changes of memory allocation with small changes in node count.
+        double adjustment;
+        if (isBetween(count, 0, 50)) {
+            adjustment = 0;
+        } else if (isBetween(count, 51, 100)) {
+            adjustment = 0.15;
+        } else if (isBetween(count, 101, 200)) {
+            adjustment = 0.3;
+        } else if (isBetween(count, 201, 300)) {
+            adjustment = 0.45;
+        } else {
+            adjustment = 0.6;
+        }
+
+        double newMemory = memory + adjustment;
+        if (count > 50) {
+            log.log(INFO, "Adjusted cluster controller memory (" + count + " content nodes): " + newMemory + " GiB");
+        }
+        return newMemory;
+    }
+
+    public static boolean isBetween(int x, int lower, int upper) {
```

Comment:
Maybe either the method name or the variable names should indicate that the value range is inclusive and not exclusive.